### ISSUE TITLE
Hi, just curious, are you able to replicate the result of the paper with the same or any settings?

### ISSUE THREAD

**Issue by yxchng** [Paralysis/sphereface#1]
I am trying to implement this as well but I am unable to replicate the result of the paper with the same settings.

**Comment by Paralysis** [comment]
Yes, exactly the same as the Caffe code they provided. I actually compared the gradient value between my implementation and their Caffe code. The reason for clamp is to make the computation stable, after all we are dividing the norm right? What if they are very close to zero...

**Comment by yxchng** [comment]
@Paralysis 
1. And do you get the same accuracy as caffe?
2. As far as I know, the clamp only matters at regions near 1 and -1, not 0/

**Comment by Paralysis** [comment]
1. Yes, I got the same performance as Caffe version. Be sure to set everything exactly the same (data preprocessing, learning rate, initialization method, etc.). 
2. We have _x*w / xlen / wlen_ right? Just in case that _xlen_ and/or _wlen_ is very close to zero. In such cases there are possibilities that result be less accurate, i.e out of the range [-1, 1].

**Comment by Paralysis** [comment]
That is a very good catch! I also noticed that when I was re-implementing the PyTorch version as well. 

1. If you check the original code in Caffe. The Xavier initialization only take the number of input channel as the parameter. However, in PyTorch, the so-called "Xavier" initializer takes both input channel and output channel as the parameter. Only the "kaiming_uniform" initializer matches with the Caffe Xavier. 

2. You are right. The first Conv layer in each block is different, i.e. Xavier initialization without bias. The others are Gaussian with bias. That's all in my PyTorch implementation. The only possible difference during training is epoch training (PyTorch) vs iteration training (Caffe). However that didn't affect the performance. 

I may upload my model some day in the future but I think it's pretty easy to implement yourself.

**Comment by yxchng** [comment]
For (2), it should be Xavier with bias and Gaussian without bias, right?

One more question, what is the corresponding decay_mult in pytorch?

I also think it should be easy but I just can't replicate. :(

**Comment by yxchng** [comment]
@Paralysis Do you use the same iteration for lr change as well? I am currently able to get the same accuracy if i train 20 epochs, with step size 10, 15, 18. However, caffe's training is about 16 epochs with 8, 12, 16 step sizes.
